sql/core/src/main/scala/org/apache/spark/sql/execution/joins/BroadcastNestedLoopJoinExec.scala

```diff
@@ -195,122 +195,103 @@ case class BroadcastNestedLoopJoinExec(
-            matched.set(i)
+          if (matchedBroadcastRows.get(i) == exists) {
+            buf += buildRows(i).copy()
           }
           i += 1
         }
-      }
-      Seq(matched).toIterator
+        sparkContext.makeRDD(buf)
     }
+  }
 
-    val matchedBroadcastRows = matchedBuildRows.fold(
-      new BitSet(relation.value.length)
-    )(_ | _)
-
-    joinType match {
-      case LeftSemi =>
-        assert(buildSide == BuildLeft)
-        val buf: CompactBuffer[InternalRow] = new CompactBuffer()
-        var i = 0
-        val rel = relation.value
-        while (i < rel.length) {
-          if (matchedBroadcastRows.get(i)) {
-            buf += rel(i).copy()
+  /**
+   * The implementation for ExistenceJoin
+   */
+  private def existenceJoin(relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
+    buildSide match {
+      case BuildRight =>
+        streamed.execute().mapPartitionsInternal { streamedIter =>
+          val buildRows = relation.value
+          val joinedRow = new JoinedRow
+
+          if (condition.isDefined) {
+            val resultRow = new GenericInternalRow(Array[Any](null))
+            streamedIter.map { row =>
+              val result = buildRows.exists(r => boundCondition(joinedRow(row, r)))
+              resultRow.setBoolean(0, result)
+              joinedRow(row, resultRow)
+            }
+          } else {
+            val resultRow = new GenericInternalRow(Array[Any](buildRows.nonEmpty))
+            streamedIter.map { row =>
+              joinedRow(row, resultRow)
+            }
           }
-          i += 1
         }
-        return sparkContext.makeRDD(buf)
-      case _: ExistenceJoin =>
+      case BuildLeft =>
+        val matchedBroadcastRows = getMatchedBroadcastRowsBitSet(streamed.execute(), relation)
         val buf: CompactBuffer[InternalRow] = new CompactBuffer()
         var i = 0
-        val rel = relation.value
-        while (i < rel.length) {
+        val buildRows = relation.value
+        while (i < buildRows.length) {
           val result = new GenericInternalRow(Array[Any](matchedBroadcastRows.get(i)))
-          buf += new JoinedRow(rel(i).copy(), result)
+          buf += new JoinedRow(buildRows(i).copy(), result)
           i += 1
         }
-        return sparkContext.makeRDD(buf)
-      case LeftAnti =>
-        val notMatched: CompactBuffer[InternalRow] = new CompactBuffer()
-        var i = 0
-        val rel = relation.value
-        while (i < rel.length) {
-          if (!matchedBroadcastRows.get(i)) {
-            notMatched += rel(i).copy()
-          }
-          i += 1
-        }
-        return sparkContext.makeRDD(notMatched)
-      case _ =>
+        sparkContext.makeRDD(buf)
     }
+  }
 
+  /**
+   * The implementation for these joins:
+   *
+   *   LeftOuter with BuildLeft
+   *   RightOuter with BuildRight
+   *   FullOuter
+   */
+  private def defaultJoin(relation: Broadcast[Array[InternalRow]]): RDD[InternalRow] = {
```

Comment:
`defaultJoin` now has outer joins only, so we cannot merge it with `outerJoin`?

Reply:
@maropu - we can, but I feel the code is too complicated as one method when combining these two.